v2/devdoc/logging.md

```diff
@@ -0,0 +1,221 @@
+
+`format` shall be a `printf`-like format string.
+
+`...` shall be the variable number of arguments required by `format`.
+
+### LOGGER_LOG
+
+```c
+#define LOGGER_LOG(log_level, log_context, format, ...) \
+    ...
+```
+
+`LOGGER_LOG` is a helper function that automatically adds the file, function and line number information to the log line.
+
+It calls `logger_log` under the hood with the arguments passed by the user while adding the file, function and line number information.
+
+## Contextual logging
+
+For each log line the library shall support specifying a context to be added to the log line.
+
+The context shall allow defining property-value pairs that are added to the event when the event is emitted.
+
+### Local stack context creation
+
+The library shall support defining a context using only stack memory, not requiring any memory allocation:
+
+```c
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_PROPERTY` or `LOG_CONTEXT_STRING_PROPERTY` entries.
+
+Example:
+
+```c
+    LOG_CONTEXT_LOCAL_DEFINE(log_context, NULL, LOG_CONTEXT_STRING_PROPERTY(property_name, "%s", MU_P_OR_NULL(prop_value)));
+```
+
+Bonus:
+
+An ideal implementation for the stack allocated contexts is also to not do any memory copy until actually needed (until they are used).
+
+That would allow writing code like below while not performing the memory copies until the logging is actually executed.
+
+```c
+void a(LOG_CONTEXT_HANDLE log_context, const char* correlation_id)
+{
+    LOG_CONTEXT_LOCAL_DEFINE(local_context, NULL, LOG_CONTEXT_STRING_PROPERTY(correlation_id, "%s", correlation_id));
+
+    if (malloc(42) == NULL)
+    {
+        logger_log(LOG_LEVEL_ERROR, local_context, "malloc failed");
+    }
+}
+```
+
+### Dynamically allocated context creation
+
+The library shall support creating a dynamically allocated context.
+
+```c
+#define LOG_CONTEXT_CREATE(log_context_handle, parent_context, ...) \
+
+#define LOG_CONTEXT_CREATE(log_context_handle) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_STRING_PROPERTY`.
+
+Example:
+
+```c
+    LOG_CONTEXT_CREATE(dynamically_allocated_log_context, NULL, LOG_CONTEXT_STRING_PROPERTY(property_name, "%s", MU_P_OR_NULL(prop_value)));
+    LOG_CONTEXT_DESTROY(dynamically_allocated_log_context);
+```
+
+### Context chaining
+
+It shall be supported to chain contexts (define a context or create a context dynamically while specifying a parent context to inherit the information from).
+
+### LOG_CONTEXT_PROPERTY
+
+`LOG_CONTEXT_PROPERTY` is a macro that allows defining non-string properties by specifying a property type, name and a value.
+
+```c
+#define LOG_CONTEXT_PROPERTY(property_type, property_name, property_value) \
+    ...
+```
+
+The following types shall be supported out of the box:
+
+- ansi_charptr
```

Comment:
>ansi_charptr [](http://example.com/codeflow?start=2&length=12)

please consider having wchar_t too

Reply:
Agree, but let's add that as a separate task.

Reply:
ID Work Item Type Title State Assigned To Effort Remaining Work Original Estimate Value Area Iteration Path Tags
15258058 Task Add wchar_t_ptr as type for v2 logging To Do           One\Custom\Azure Messaging Store\Iteration 271